### PR TITLE
Include full SSL Cert Chain in deployed SSL cert for Nginx

### PR DESCRIPTION
**Pull request by hnrkndrssn**
Fixes OctopusDeploy/Issues#8360

### CODE REVIEW

File: source/Calamari.Tests/Fixtures/Nginx/Approved/NginxFixture.SetupReverseProxyWithSslUsingCertificateVariableSite.crt.approved.txt [hunk 1]
```diff
@@ -0,0 +1,63 @@
+-----BEGIN CERTIFICATE-----
```
Comment (hnrkndrssn) [inline source/Calamari.Tests/Fixtures/Nginx/Approved/NginxFixture.SetupReverseProxyWithSslUsingCertificateVariableSite.crt.approved.txt:1]: This cert is just a test certificate (that is expired) and doesn't expose anything secret

File: source/Calamari.Tests/Fixtures/Nginx/NginxFixture.cs [hunk 2]
```diff
@@ -285,5 +294,33 @@ public void TestLocationsUnsuitableForFilenames()
             // Ensure the filename matches the expected format.
             Assert.IsTrue(Directory.GetFiles(Path.Combine(tempDirectory, "conf"))[0].EndsWith($"{nginxServer.VirtualServerName}.conf"));
         }
+
+        (string, string, string) GetCertificateDetails()
+        {
+#if NETCORE
+            var chainCertFilePath = TestEnvironment.GetTestPath("Helpers", "Certificates", "SampleCertificateFiles", "3-cert-chain.pfx");
```
Comment (hnrkndrssn) [inline source/Calamari.Tests/Fixtures/Nginx/NginxFixture.cs:301]: `net4x` and `mono` does not support  `X509Certificate2Collection` and other stuff done here to get data from our test cert.

File: source/Calamari/Integration/Nginx/NginxServer.cs [hunk 3]
```diff
@@ -116,6 +117,10 @@ public NginxServer WithServerBindings(IEnumerable<Binding> bindings, IDictionary
             }
 
             return this;
+
+            string GetCertificateWithChainIfPresent(string certificatePem, string chainPem)
+                => $@"{certificatePem}
```
Comment (hnrkndrssn) [inline source/Calamari/Integration/Nginx/NginxServer.cs:122]: Main certificate *must* come first in the `.crt` file

File: source/Calamari.Tests/Fixtures/Nginx/NginxFixture.cs [hunk 4]
```diff
@@ -285,5 +294,33 @@ public void TestLocationsUnsuitableForFilenames()
...
+            {
+                // if the cert has no extension or the cert has a basic constraints extension,
+                // and the certificate authority is true then this is a chain certificate
+                certificateChainPem += $@"{new string(PemEncoding.Write("CERTIFICATE", cert.RawData))}
+";
+            }
+
+            return (certificatePem, certificatePrivateKeyPem, certificateChainPem);
+#else
+            return ("", "", "");
```
Comment (benPearce1) [inline source/Calamari.Tests/Fixtures/Nginx/NginxFixture.cs:322]: nit: prefer `string.Empty` over `""`
There is no technical/perf reason for this, just readability and clarity of intent.

File: source/Calamari/Integration/Nginx/NginxServer.cs [hunk 5]
```diff
@@ -116,6 +117,10 @@ public NginxServer WithServerBindings(IEnumerable<Binding> bindings, IDictionary
             }
 
             return this;
+
+            string GetCertificateWithChainIfPresent(string certificatePem, string chainPem)
```
Comment (benPearce1) [inline source/Calamari/Integration/Nginx/NginxServer.cs:121]: I am unsure what this is actually supposed to return?

Comment (hnrkndrssn) [reply source/Calamari/Integration/Nginx/NginxServer.cs:121]: It returns the server certificate and any chained certificates combined into a new-line separated string.

**Review by benPearce1 — APPROVED** [review]
LGTM, not able to test locally

File: source/Calamari/Integration/Nginx/NginxServer.cs (hunk 5 above)
Comment (hnrkndrssn) [reply source/Calamari/Integration/Nginx/NginxServer.cs:121]: Would `GetServerCertificateCombinedWithAnyChainedCertificates` make it easier to grok?

Comment (hnrkndrssn) [reply source/Calamari/Integration/Nginx/NginxServer.cs:121]: Or maybe `CombineServerCertificateWithAnyChainedCertificates`

Comment (benPearce1) [reply source/Calamari/Integration/Nginx/NginxServer.cs:121]: no, what you have is good enough, I just wasn't sure what exactly it was writing out, at first I thought it was a file path.

**Pull request merged**